RF24.cpp

```diff
@@ -985,6 +1016,608 @@ void RF24::sprintfPrettyDetails(char *debugging_information)
+    BitManip::packValue(getPayloadSize(), 8, encoded_details, &encoded_details_index, &bit_index);
+    // 11 (read_register(SETUP_RETR) >> ARD)
+    temp_8_bit = (read_register(SETUP_RETR) >> ARD);
+    BitManip::packValue(temp_8_bit, 8, encoded_details, &encoded_details_index, &bit_index);
+    // 12 (read_register(SETUP_RETR) & 0x0F)
+    temp_8_bit = (read_register(SETUP_RETR) & 0x0F);
+    BitManip::packValue(temp_8_bit, 8, encoded_details, &encoded_details_index, &bit_index);
+    // 13 (read_register(OBSERVE_TX) >> 4)
+    temp_8_bit = (read_register(OBSERVE_TX) >> 4);
+    BitManip::packValue(temp_8_bit, 8, encoded_details, &encoded_details_index, &bit_index);
+    // 14 (read_register(OBSERVE_TX) & 0x0F)
+    temp_8_bit = (read_register(OBSERVE_TX) & 0x0F);
+    BitManip::packValue(temp_8_bit, 8, encoded_details, &encoded_details_index, &bit_index);
+    // 15 static_cast<bool>(read_register(FEATURE) & _BV(EN_DYN_ACK))
+    temp_bool = static_cast<bool>(read_register(FEATURE) & _BV(EN_DYN_ACK));
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
+    // 16 static_cast<bool>(read_register(FEATURE) & _BV(EN_ACK_PAY))
+    temp_bool = static_cast<bool>(read_register(FEATURE) & _BV(EN_ACK_PAY));
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
+    // 17 ((read_register(DYNPD) && (read_register(FEATURE) &_BV(EN_DPL))) * 1)
+    temp_8_bit = ((read_register(DYNPD) && (read_register(FEATURE) & _BV(EN_DPL))) * 1);
+    BitManip::packValue(temp_8_bit, 8, encoded_details, &encoded_details_index, &bit_index);
+    // 18 (static_cast<bool>(temp_8_bit & _BV(ENAA_P5)) + 48)
+    temp_8_bit = read_register(EN_AA);
+    temp_bool = (static_cast<bool>(temp_8_bit & _BV(ENAA_P5)) + 48);
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
+    // 19 (static_cast<bool>(temp_8_bit & _BV(ENAA_P4)) + 48)
+    temp_bool = (static_cast<bool>(temp_8_bit & _BV(ENAA_P4)) + 48);
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
+    // 20 (static_cast<bool>(temp_8_bit & _BV(ENAA_P3)) + 48)
+    temp_bool = (static_cast<bool>(temp_8_bit & _BV(ENAA_P3)) + 48);
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
+    // 21 (static_cast<bool>(temp_8_bit & _BV(ENAA_P2)) + 48)
+    temp_bool = (static_cast<bool>(temp_8_bit & _BV(ENAA_P2)) + 48);
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
+    // 22 (static_cast<bool>(temp_8_bit & _BV(ENAA_P1)) + 48)
+    temp_bool = (static_cast<bool>(temp_8_bit & _BV(ENAA_P1)) + 48);
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
+    // 23 (static_cast<bool>(temp_8_bit & _BV(ENAA_P0)) + 48)
+    temp_bool = (static_cast<bool>(temp_8_bit & _BV(ENAA_P0)) + 48);
+    BitManip::packBool(temp_bool, encoded_details, &encoded_details_index, &bit_index);
```

Comment:
Yes, I will work on that later those were for me.  It will.

> ### 1st review
> 
> I mostly focused on conceptual stuff here. For the next review, I intend to focus on better code reuse since... well ![image](https://user-images.githubusercontent.com/14963867/151109180-ac38ecd4-e306-4dda-ace7-807299128cdd.png) We need to work on this.
> 
> Also, I thought the original idea for these changes was to compliment `sprintfPrettyDetails()` (& friends), but I don't see that code taking advantage of these changes. Is my timing a bit premature?

Yes, I want to get rid of the sprintf in encodedRadioDetails, I think we should change the function declaration to have a default argument sprintfPrettyDetails(debugging_information, encoded_array=0).  That would get rid of quite a few lines, and increase maintainability because there aren't two format_string doing similar things, it wouldn't break anyone's existing project who decided to use the new feature.  I think that's a good approach, what do you think? There are a lot of other things I need to take care of that you pointed out so once I'm through with that I think we should be in a good place to change the output behavior once it's verified correct.

Reply:
I think changing `decodeDetails()` into an overload of `sprintfPrettyDetails()` that takes 2 args instead of 1 would be a better approach.

I imagine it like so
```cpp
void sprintfPrettyDetails(char* debug_info) {
    // Create buffer using encodeDetails()
    uint8_t packedDets[38];
    encodeDetails(&packedDets);

    // Pass the buffer to the overloaded function
    sprintfPrettyDetails(debug_info, &packedDets);
}
```

Reply:
> I think changing `decodeDetails()` into an overload of `sprintfPrettyDetails()` that takes 2 args instead of 1 would be a better approach.
> 
> I imagine it like so
> 
> ```c++
> void sprintfPrettyDetails(char* debug_info) {
>     // Create buffer using encodeDetails()
>     uint8_t packedDets[38];
>     encodeDetails(&packedDets);
> 
>     // Pass the buffer to the overloaded function
>     sprintfPrettyDetails(debug_info, &packedDets);
> }
> ```

`
sprintfPrettyDetails(char *debugging_information, uint8_t *encoded_details=0)
{
 if(encoded_details != 0)
  {
   //use encoded details for output
  }
 else
 {
  //read registers as normal
 }
 //output
}
decodeRadioDetails(char *debugging_information, uint8_t *encoded_details)
{
 sprintfPrettyDetails(debugging_information, encoded_details);
}
`

is what I was thinking of

Reply:
I'm sorry for the bad code comment lol, I'm learning this as fast as I can.

Reply:
so really, it ends up just being a wrapper for sprintfPrettyDetails.

Reply:
Github uses [markdown syntax](https://docs.github.com/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax). Use ["fenced" code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) to preserve whitespace.

Reply:
> `if(encoded_details != 0)`

This is a classic NULL vs 0 problem. I was trying to avoid this by using function overloads.